microsoft365defender_connector.py

```diff
@@ -1150,6 +1150,164 @@ def _handle_update_alert(self, param):
+                try:
+                    self._ingest_artifacts_new(artifacts, name=incident["displayName"], key=incident["id"])
+                except Exception as e:
+                    self.debug_print("Error occurred while saving artifacts for incidents. Error: {}".format(str(e)))
+
+            if self.is_poll_now():
+                break
+
+            if incident_list:
+                if DEFENDER_JSON_LAST_MODIFIED not in incident_list[-1]:
+                    return action_result.set_status(phantom.APP_ERROR, "Could not extract {} from latest ingested "
+                                                                       "incident.".format(DEFENDER_JSON_LAST_MODIFIED))
+
+                self._state[STATE_LAST_TIME] = incident_list[-1].get(DEFENDER_JSON_LAST_MODIFIED)
+                self.save_state(self._state)
+
+            offset += incident_left
+            incident_left = self.duplicate_container
+            self.duplicate_container = 0
+
+        return action_result.set_status(phantom.APP_SUCCESS)
+
+    def _ingest_artifacts_new(self, artifacts, name, key):
+        """Save the artifacts into the given container ID(cid) and if not given create new container with given key(name).
+        Parameters:
+            :param artifacts: list of artifacts of IoCs or incidents results
+            :param name: name of the container in which data will be ingested
+            :param key: source ID of the container in which data will be ingested
+        Returns:
+            :return: status(phantom.APP_SUCCESS/phantom.APP_ERROR), message, cid(container_id)
+        """
+        container = {
+            "name": name,
+            "description": 'incident ingested using MS Defender API',
+            "source_data_identifier": key
+        }
+
+        ret_val, message, cid = self.save_container(container)
+        if phantom.is_fail(ret_val):
+            self.debug_print("Error occurred while creating container, reason: {}".format(message))
+            return
+
+        self.debug_print("save_container (with artifacts) returns, value: {}, reason: {}, id: {}".format(ret_val, message, cid))
+        if message in "Duplicate container found":
+            self.duplicate_container += 1
+            self.debug_print("Duplicate container count: {}".format(self.duplicate_container))
+
+        for artifact in artifacts:
+            artifact['container_id'] = cid
+        ret_val, message, _ = self.save_artifacts(artifacts)
+
+        self.debug_print("save_artifacts returns, value: {}, reason: {}".format(ret_val, message))
+
+    @staticmethod
+    def _create_alert_artifacts(alert):
+
+        return {'label': 'alert', 'name': alert.get('title'), 'source_data_identifier': alert.get('id'),
+                'data': alert, 'cef': alert}
+
+    @staticmethod
+    def _create_incident_artifacts(incident):
+        return {'label': 'incident', 'name': 'incident Artifact', 'source_data_identifier': incident.get('id'),
```

Comment:
I'd format this like the one above

Reply:
actually, pycharm reformats it like this, however, I will update it

Reply:
Done